### ISSUE TITLE
Automatic conversion from float to int???

### ISSUE THREAD

**Issue by nandoflorestan** [samuelcolvin/pydantic#360]
# Bug

For bugs/questions:
* OS: **Any**
* Python version `import sys; print(sys.version)`: **3.7**
* Pydantic version `import pydantic; print(pydantic.VERSION)`: **0.18.1**

You got a misfeature there, buddy.  Automatic conversion (by default!) from float to int is a terrible idea:

```py
from pydantic import PositiveInt, BaseModel

class MySchema(BaseModel): 
    int_only: PositiveInt

MySchema(int_only=3.1415)
# <MySchema int_only=3>
```

I am working with 2 separate codebases that communicate with each other.  One of these apps could have the wrong idea about the type of a variable, it would be great if the validation framework caught this. I wanted to use Pydantic to validate the data.  How can I circumvent this misfeature?  (Please add the answer to the documentation.)

**Comment by samuelcolvin** [comment]
> You got a misfeature there, buddy. Automatic conversion (by default!) from float to int is a terrible idea

Wow, what a great way of not getting help.

I try really hard to understand peoples problems and always give constructive answers, but here I'm really tempted to just click close. 

Do you always start conversations with volunteers where you would like help by calling them "buddy", then telling them they have "terrible idea"s? Great idea buddy.

This is by considered design, but you can change the behaviour using validators with `pre=True`. 

Or you can use another library. Not a problem with me.

**Comment by samuelcolvin** [comment]
Ref. #284

**Comment by nandoflorestan** [comment]
Dear Samuel,

I apologize.  You are absolutely right that I was arrogant.  It's nothing personal.  I repeat this mistake nearly every week and then I resolve to watch it and then suddenly it happens again.  I seem unable to shake it, but I must not give up.  I am not being sarcastic, I have struggled with this all my life.  Therapy had no power to shadow my dad's example.  If it means anything, no offense was intended, but I spoke as if I already knew you, trying to be a little funny, and that's the way it always happens.

Again, I am sorry and I hope you can forgive me.  I did not really mean to talk down to you, although it totally appears that way.  I truly admire your work in this library (one that I couldn't have written), I think it is awesome, and this is why I am going to insist.

I still think that the default should be changed -- we should not build a validation library that declines to validate types by default.  (Python is a strongly typed, if dynamically typed, language.)  In the case I brought to you, 3.14 is being converted to 3 by default, which is information loss, happening exactly when we wanted to prevent information loss.

At a minimum, a configuration option should be added to control this behavior.

I guess that's the best case I can make for it.

Best regards, Nando.

**Comment by samuelcolvin** [comment]
Please read #284.

Also python is not a strictly typed language, it's duck typed by default. Look at bool behaviour, it's as unstrict as unimaginable.

**Comment by nandoflorestan** [comment]
We can simplify this discussion.  I do not mean to discuss Python or its booleans.  I am just arguing that allowing 3.14 to become 3 is data loss.  I am saying that information loss is a bad default, and that we should do something about this.  We can restrict the discussion to the float to int case.

**Comment by samuelcolvin** [comment]
I'm not interested in Changing the behaviour even to make it configurable, as explained in #284 that would be a massive amount of work and would likely cause large amounts of partially duplicated code, or performance degradation, or both.

You can however achieve this behaviour with custom validators with `pre=True`.

I'll add a section to the docs explaining that pydantic is a parsing library not a validation library and therefore makes promises about the form of output models, not the input data. I'll also remove references to validation which could be argued to be inaccurate.

**Comment by samuelcolvin** [comment]
But

> I apologize...

Thanks a lot for this, admitting ones mistakes is never easy and your humility is much appreciated.

**Comment by four43** [comment]
This still open bug makes this entire library not usable IMO. Defaulting to crazy magic type corrosion is a terrible idea.

**Comment by dgasmith** [comment]
Coercion in some places such as JSON types to Python sets, tuples, etc makes a lot of sense and greatly eases quite a bit of programming. Not coercing standard python allowed types would make things quite a bit more complex. As mentioned, this is quite easy to do yourself:

```python
from pydantic import BaseModel

class StrictInt(int):
    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v):
        if isinstance(v, float):
            raise TypeError("Will not coerce float to int")
        return v

class Model(BaseModel):
    v: StrictInt

print(Model(v=5))
print(Model(v=5.0))
```

The validator can be extended to allow `5.0` but not `5.1` for example based on tolerances. As a base library this is the functionality that you want.

**Comment by samuelcolvin** [comment]
@four43, thanks for your feedback. Perhaps this is "crazy magic", but it's pythonic crazy magic:

```py
>>> int('123')
123
```

Please could you give a practical example of where this approach "makes this entire library not usable (sic)"?

I'd also really like to understand why you've given thumbs down to my "thanks" comment [above](https://github.com/samuelcolvin/pydantic/issues/360#issuecomment-455819182) seems fairly innocuous to me...

**Comment by jbednar** [comment]
If anyone is interested in a somewhat similar library that _is_ designed for validation, you might consider [param.holoviz.org](param.holoviz.org). Unlike pydantic, which is clearly labeled as not being a validation library, Param is designed to give feedback to the caller about inappropriate inputs, raising exceptions rather than coercing data into the expected types.  Param is explicitly designed for the case outlined by the OP, i.e. working with separate codebases that communicate with each other and need to agree very clearly and explicitly what is and is not acceptable input.

**Comment by four43** [comment]
I'm trying to remember my work around for this, I think I had to use `Strict` prefixes for everything. It's just a bummer it's not explicit by default and relies on the dev to ensure ints/floats don't get clobbered.

**Comment by bluenote10** [comment]
> Perhaps this is "crazy magic", but it's pythonic crazy magic:

But Python only takes it to a point were it is safe:

![grafik](https://user-images.githubusercontent.com/3620703/128723094-e846cb5d-1845-4055-85a3-a61f56e144b2.png)

Going beyond Python's default coercion behaviour is indeed opening the doors for hard-to-find bugs, and is actually rather unexpected in particular as an experienced Python developer.

**Comment by kingcowman** [comment]
> > Perhaps this is "crazy magic", but it's pythonic crazy magic:
> 
> But Python only takes it to a point were it is safe:
> 
> ![grafik](https://user-images.githubusercontent.com/3620703/128723094-e846cb5d-1845-4055-85a3-a61f56e144b2.png)
> 
> Going beyond Python's default coercion behaviour is indeed opening the doors for hard-to-find bugs, and is actually rather unexpected in particular as an experienced Python developer.

I think there's a miscommunication here.

Pydantic produces an exception in the same case, it does not go beyond default type coercion.

![image](https://user-images.githubusercontent.com/8564453/134412490-f920ee76-c985-438d-b4c6-394b5e724c79.png)

![image](https://user-images.githubusercontent.com/8564453/134412581-15135051-64d6-4412-8e43-d2548f52831f.png)

**Comment by bluenote10** [comment]
My point was that there is an inconsistency here, also highlighted in your example:
- `foo="1.2"` gives a "not a valid integer" error.
- `foo=1.2` gives no error, and secretively loses information.

Looking at it from a "pedantic" perspective: `"1.2`" isn't a valid integer and neither is `1.2`. This can cause bugs, because (1) the inconsistency and (2) the possibility of losing information. The behavior caught me by surprise especially because it contradicts the library's name.

**Comment by sampathpalanivel** [comment]
Hi,
Int type  also accepting bool
{
  "age": true
}

output:
{
  "age": 1
}
is this correct?

**Comment by biermeester** [comment]
Just for reference:

It seems that nowadays you can define a value to strictly be an integer, e.g.:

```voyage_id: conint(strict=True)```

**Comment by Danstiv** [comment]
Why doesn't latest pydantic (2.3.0) convert float to int?

```
class TestModel(BaseModel):
    n: int

TestModel(n=3.14)  # Gives a validation error
```
The documentation says that pydantic tries to coerce the data, and uses int(v) for int.
int(3.14) works fine, and in 1.x it seems to have worked. What am I missing?

**Comment by nandoflorestan** [comment]
@Danstiv
You are missing [my comment in this issue](https://github.com/pydantic/pydantic/issues/360#issuecomment-455772172) which considers it data loss when 3.14 becomes 3 without an explicit line of code to do so.  When you say "works fine", you are missing this point, in fact inverting it.  For a data validation library to "work fine", the first thing that must be respected are data types such as int and float.

**Comment by tadasgedgaudas** [comment]
How do you guys convert float to int now? Do you use some validators? I cannot find a clean way to do that now

**Comment by bensaufley** [comment]
I think the original complaint here was valid, but the prior reasoning ("behave as Python behaves") had some value as well. We recently upgraded to Pydantic 2 and encountered a host of places where we had been silently converting floats with fractions to ints, now loudly failing.  We've chosen to deal with those at the source, but during investigation, after trying @biermeester's suggestion above I found that `conint` is `!!! warning "Discouraged"` in favor of `Annotation` but I can't figure out how to make that work: `Annotation[int, Field(strict=False)]` still raises if the float has a fraction. I'm wondering what the suggested solution is here, today, if we want to be up-to-date but continue with Python-like coercion?
